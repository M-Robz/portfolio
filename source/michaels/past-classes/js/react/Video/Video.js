import React, { useContext, useRef } from 'react';

// Custom modules
const extractYtVideoId = require('../../modules/extractYtVideoId');

// Hooks
import useEffectOnUpdate from '../hooks/useEffectOnUpdate';

// Components
import Skeleton from 'react-loading-skeleton';

// Store
import GlobalContext from '../_template/store/global-context';
import LoadingContext from '../store/loading-context';

/**
 * An embedded YouTube video of a class, with additional information beneath.
 * @param {Object} props
 * @param {Object} props.data - Data for the video.
 *
 * Note: Using PureComponent avoids unnecessary re-rendering, but breaks the
 * lazyloading solution.
 */
export default function Video({ data }) {
    const iframeRef = useRef(null);
    const { region, locale } = useContext(GlobalContext);
    const loading = useContext(LoadingContext);

    /*
     * On every re-render, restore the `mel-js-lazy` class to the iframe so it
     * can be loaded again. Without this fix, the video would disappear and not
     * be reloaded when the component re-rendered.
     */
    useEffectOnUpdate(() => {
        if (iframeRef.current) {
            // Note: Using add/remove because IE doesn't support
            // `DOMTokenList.replace`
            iframeRef.current.classList.remove('lazyloaded');
            iframeRef.current.classList.add('mel-js-lazy');
        }
    });

    if (loading) {

        return (
            <div className="c-Video">
                <div className="c-Video__video mel-c-video">
                    <div className="mel-c-video__fluid">
                        <Skeleton className="c-Video__skeleton" />
                    </div>
                </div>
                <div className="c-Video__text">
                    <h3 className="c-Video__heading">
                        <Skeleton count={3} />
                    </h3>
                    <p className="c-Video__content">
                        <Skeleton width="5em" />
                    </p>
                </div>
            </div>
        );

    } else {

        // Read data
        const videoId = extractYtVideoId(data['Video ID']),
              title = data[`Title ${ locale.toUpperCase() }`],
              host = data['Hosted by'],
              displayDate = region.isQuebec
                  ? _formatQUDate(data['Class Date'])
                  : data['Class Date']
        ;

        return (
            <div className="c-Video">
                <div className="c-Video__video mel-c-video">
                    <div className="mel-c-video__fluid">
                        <iframe
                            className="mel-c-video__player mel-js-lazy"
                            data-src={`https://www.youtube.com/embed/${videoId}?rel=0&modestbranding=1`}
                            allowFullScreen
                            width="412"
                            height="232"
                            title="Video"
                            ref={iframeRef}>
                        </iframe>
                    </div>
                </div>
                <div className="c-Video__text">
                    <h3 className="c-Video__heading">
                        {title}
                        { host && ' | ' }
                        { host &&
                            <span className="c-Video__host">{host}</span>
                        }
                    </h3>
                    { displayDate &&
                        <p className="c-Video__content">{displayDate}</p>
                    }
                </div>
            </div>
        );
    }
}

/**
 * Format a date for Quebec.
 * @private
 * @param {string} date - Date in the format M/D/Y
 * @returns {string} Date in the format D/M/Y
 */
function _formatQUDate(date) {
    const arr = date.split('/');
    return arr.length === 3 ? `${arr[1]}/${arr[0]}/${arr[2]}` : '';
}
