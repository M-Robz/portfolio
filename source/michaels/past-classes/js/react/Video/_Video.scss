.c-Video {
    padding: 0 0 1.5rem;
    text-align: left;

    @media (min-width: $medium-min) {
        padding: 0 0.5rem 1.8rem;
    }
}

.c-Video__skeleton {
    position: absolute;
    height: 100%;
}

.c-Video__video {
    margin-bottom: 0.5rem;

    @media (min-width: $large-min) {
        margin-bottom: 0.8rem;
    }
}

.c-Video__heading {
    margin: 0 0 0.2rem;
    font-size: 1em;
    line-height: 1.3;
    font-weight: normal;
}

// For two-column layout on mobile
.mel-string.mel-string-2-up .c-Video__heading {

    @media (max-width: $xs-max) {
        font-size: 0.9em;
    }
}

.c-Video__host {
    font-weight: bold;
}

.c-Video__content {
    margin: 0;
    font-size: 0.9em;
    line-height: 1.3;
    color: #444;
}
