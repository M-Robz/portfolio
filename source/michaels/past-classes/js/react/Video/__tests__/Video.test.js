import React from 'react';
import renderer from 'react-test-renderer';
import GlobalContext from '../../_template/store/global-context';
import LoadingContext from '../../store/loading-context';
import Video from '../Video';

const GLOBAL_STATE = {
    region: {
        isCanada: false,
        isQuebec: false
    },
    locale: 'us'
};

const DATA = {
    'Video ID': 'https://www.youtube.com/watch?v=7bLxzojWrfA',
    'Title US': '3 Hour Throw Crochet with no tools!',
    'Hosted by': 'Yarnspirations',
    'Class Date': '6/20/2022'
};

it('renders skeletons while loading', () => {
    // TODO: better way to test?
    const tree = renderer.create(
        <GlobalContext.Provider value={GLOBAL_STATE}>
            <LoadingContext.Provider value={true}>
                <Video />
            </LoadingContext.Provider>
        </GlobalContext.Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});

it('renders correctly once loaded', () => {
    const tree = renderer.create(
        <GlobalContext.Provider value={GLOBAL_STATE}>
            <LoadingContext.Provider value={false}>
                <Video data={DATA} />
            </LoadingContext.Provider>
        </GlobalContext.Provider>
    ).toJSON();
    expect(tree).toMatchSnapshot();
});
