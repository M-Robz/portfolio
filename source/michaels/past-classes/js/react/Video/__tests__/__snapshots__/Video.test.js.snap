// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders correctly once loaded 1`] = `
<div
  className="c-Video"
>
  <div
    className="c-Video__video mel-c-video"
  >
    <div
      className="mel-c-video__fluid"
    >
      <iframe
        allowFullScreen={true}
        className="mel-c-video__player mel-js-lazy"
        data-src="https://www.youtube.com/embed/7bLxzojWrfA?rel=0&modestbranding=1"
        height="232"
        title="Video"
        width="412"
      />
    </div>
  </div>
  <div
    className="c-Video__text"
  >
    <h3
      className="c-Video__heading"
    >
      3 Hour Throw Crochet with no tools!
       | 
      <span
        className="c-Video__host"
      >
        Yarnspirations
      </span>
    </h3>
    <p
      className="c-Video__content"
    >
      6/20/2022
    </p>
  </div>
</div>
`;

exports[`renders skeletons while loading 1`] = `
<div
  className="c-Video"
>
  <div
    className="c-Video__video mel-c-video"
  >
    <div
      className="mel-c-video__fluid"
    >
      <span
        aria-busy={true}
        aria-live="polite"
      >
        <span
          className="react-loading-skeleton c-Video__skeleton"
          style={Object {}}
        >
          ‌
        </span>
        <br />
      </span>
    </div>
  </div>
  <div
    className="c-Video__text"
  >
    <h3
      className="c-Video__heading"
    >
      <span
        aria-busy={true}
        aria-live="polite"
      >
        <span
          className="react-loading-skeleton"
          style={Object {}}
        >
          ‌
        </span>
        <br />
        <span
          className="react-loading-skeleton"
          style={Object {}}
        >
          ‌
        </span>
        <br />
        <span
          className="react-loading-skeleton"
          style={Object {}}
        >
          ‌
        </span>
        <br />
      </span>
    </h3>
    <p
      className="c-Video__content"
    >
      <span
        aria-busy={true}
        aria-live="polite"
      >
        <span
          className="react-loading-skeleton"
          style={
            Object {
              "width": "5em",
            }
          }
        >
          ‌
        </span>
        <br />
      </span>
    </p>
  </div>
</div>
`;
